packages/ruleset/src/functions/schema-naming-convention.js

```diff
@@ -0,0 +1,474 @@
+ * the request body to the sub-property, there will be an entry mapping from
+ * the location of the schema in components. Here are example entries:
+ * - paths./v1/things.post.requestBody.content.application/json.schema:
+ *     components.schemas.ThingPrototype
+ * - components.schemas.ThingPrototype.properties.data:
+ *     components.schemas.DataObject
+ *
+ * The purpose of this function is to take the fully resolved path (in the
+ * example above, something like 'paths./v1/things.post.requestBody.content.
+ * application/json.schema.properties.data') and un-resolve enough that it
+ * will match the format Spectral uses in nodes (e.g. 'components.schemas.
+ * ThingPrototype.properties.data').
+ *
+ * @param {string} path - the resolved JSON path to a schema, as a dot-separated string
+ * @param {object} pathToReferencesMap - the graph nodes map from Spectral, which has
+ *                                       been sanitzed for our purposes. it maps referenced
+ *                                       schema locations to their reference location.
+ * @returns {string} - the name of the referenced schema at a given path, or undefined
+ */
+function getSchemaNameAtPath(path, pathToReferencesMap) {
+  if (!path && typeof path !== 'string') {
+    return;
+  }
+
+  // resolve path to reference - add comments explaining
+  // Build the path, replacing each path that resolves to a reference with the
+  // referenced path in order to match the expected format in the
+  // pathToReferencesMap (which comes from graph nodes that Spectral gives us).
+  // See the function documentation above for more info.
+  let pathBuilder = '';
+  for (const pathSegment of path.split('.')) {
+    pathBuilder += `${pathSegment}`;
+    const schemaReference = pathToReferencesMap[pathBuilder];
+
+    // If it is the last time through the loop, we should definitely
+    // find a schema reference - but we don't want to throw away the
+    // path. We'll find the reference again at the end of this function.
+    if (schemaReference && pathSegment !== path.split('.').at(-1)) {
+      pathBuilder = schemaReference;
+    }
+
+    pathBuilder += '.';
+  }
+
+  // We lazily add a period character at the end of every segment
+  // during the loop. Remove the final period here.
+  pathBuilder = pathBuilder.slice(0, -1);
+
+  if (path !== pathBuilder) {
+    logger.debug(`${ruleId}: resolved path to be ${pathBuilder}`);
+  }
+
+  return getSchemaNameFromReference(pathToReferencesMap[pathBuilder]);
+}
+
+/**
+ * Takes a path to a referenced schema (as a string) and extracts the last element,
+ * which will be the name of the schema.
+ *
+ * @param {string} reference - the ref value as a dot-separated string (e.g. 'components.schemas.Thing')
+ * @returns {string} - the name of the schema (e.g. 'Thing')
+ */
+function getSchemaNameFromReference(reference) {
+  if (!reference && typeof reference !== 'string') {
```

Comment:
Should this also use || instead of &&?
```suggestion
  if (!reference || typeof reference !== 'string') {
```

Reply:
Yes again 🙂